### PR TITLE
Add support for reading GRIB files

### PR DESCRIPTION
**Pull request by tcarion**
See #282

### CODE REVIEW

**Review by rafaqz — COMMENTED** [review]
A few comments. Otherwise I guess you just have to keep working through things until your tests work.

Probably doing this you will duplicate a lot of code. We can move some of these things to generic utils methods that can be shared accross backends.

**Comment by tcarion** [comment]
I've created a new package [GRIBDataset](https://github.com/tcarion/GRIBDatasets.jl) that is, IMHO, easier to extend than CfGRIB.jl and that avoids re-indexing the GRIB file when we have multiple layers, making it faster to build as `RasterStack`.

File: src/sources/grib.jl [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        else
+            span, sampling = Irregular((extrema(dates)...)), Points()
+        end
+        return _cfglookup(D, index, order, span, sampling, metadata, crs, mappedcrs)
+    # elseif eltype(index) <: Dates.AbstractTime
+    #     span, sampling = _cfgperiod(index, metadata)
+    #     return _cfglookup(D, index, order, span, sampling, metadata, crs, mappedcrs)
+    else
+        # Use the same function as for NetCDF files
+        span, sampling = _ncdspan(index, order), Points()
```
[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/grib.jl:173]: What about creating a `sources/common.jl` where I put the methods that can be shared between `sources/ncdatasets.jl` and `sources/grib` ?

File: src/sources/grib.jl [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    "level" => Z,
+    "vertical" => Z,
+    "x" => X,
+    "y" => Y,
+    "z" => Z,
+    "band" => Band,
+)
+
+haslayers(::Type{GRIBfile}) = true
+defaultcrs(::Type{GRIBfile}) = EPSG(4326)
```
[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/grib.jl:15]: I admit I just copied from `ncdasasets.jl` without really understanding, just hoping to break as less things as possible :relaxed: I don't really about CRS with GRIB

File: src/sources/grib.jl [hunk 3]
```diff
@@ -0,0 +1,223 @@
+const CFG = CfGRIB
+
+const GRIB_DIMMAP = Dict(
```
[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/grib.jl:3]: I made it more specific in the last commit. I think I can't take the same as for netcdf since the relevant `time` dimension is usually `valid_time` in GRIB (as far as I know). Or maybe I should just transform the dimension name from `valid_time` to `time` in GRIBDatasets.jl, and then we could use the same dimension mapping

File: src/sources/grib.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (rafaqz) [reply src/sources/grib.jl:173]: Good idea

File: src/sources/grib.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (rafaqz) [reply src/sources/grib.jl:15]: Hah yeah there are corners of all the raster data specs I don't care about either.

But if you don't implement it correctly someone will post an issue here one day that forces me to fix it instead...

File: src/sources/grib.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (rafaqz) [reply src/sources/grib.jl:3]: Its ok to keep the standard names.

File: src/sources/grib.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/grib.jl:15]: Sorry for the typo, I didn't mean that I don't care, but that I don't know about CRS with GRIB :wink: 
I did some research and it happens that knowing the CRS from the GRIB messages is not straightforward. Reading the shapeOfTheEarth key and connecting it to [this table](https://apps.ecmwf.int/codes/grib/format/grib2/ctables/3/2) can give information about the reference system. Then I guess it could be translated to standard CRS (with PROJ4?), but I'm really not familiar with all that. I guess that for the moment we could do `defaultcrs(::Type{GRIBfile}) = nothing` and leave it to the user to determine what CRS he's dealing with?

**Comment by rafaqz** [comment]
Sorry I forgot to actually post my few review comments!  

Probably need to rebase and accept the new changes to ncdatasets.jl

**Comment by tcarion** [comment]
No problem, as you may observe, I'm not really diligent either :relaxed: 

I rebased and added some tests. It could be better, but I'll have a closer look when I have a bit more time

File: test/sources/grib.jl [hunk 4]
```diff
@@ -0,0 +1,219 @@
+using Rasters, Test, GRIBDatasets
+using Rasters: FileArray, FileStack, NCDfile, crs
+
+test_files = [
+    "era5-levels-members.grib", # OK
+    "fields_with_missing_values.grib", # OK
+    "forecast_monthly_ukmo.grib", # Failing because because there are multiple steps
```
[the remark this answers is not in the document]
Comment (tcarion) [reply test/sources/grib.jl:9]: This dataset is more tricky to handle because it has multiple forecasts with different time steps, making it tricky to calculate the valid time. But that's actually a problem that must be solved in GRIBDatasets (I'll have a look at it). We can just ignore that for the moment. 

While using GRIBDatasets and Rasters for GRIB files, I also encountered other edge cases, but again I think this can be handled later on.

File: test/sources/grib.jl [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    "lambert_grid.grib", # longitude and latitude are 2D variables. Don't know what to do in this case
```
[the remark this answers is not in the document]
Comment (tcarion) [reply test/sources/grib.jl:10]: Alright, it seems pretty obscure to me right now but I'll have a look :relieved:

File: test/sources/grib.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (rafaqz) [reply test/sources/grib.jl:9]: Absolutely, making things mostly work and fixing as we go is mostly how Rasters.jl has developed. Working on 1% edge cases when 95% workflows aren't completed mediocre is not a good use of our time.

**Comment by tcarion** [comment]
I rebased the grib branch with main, taking into account the modifications that happened to `ncdatasets.jl` in the meantime. The local tests I performed show that these commits seem to break `hasdim` on rasters created from netcdf:

```julia
julia> slice(ncarray, Z)
ERROR: BoundsError: attempt to access 0-element Vector{Union{}} at index [1]
Stacktrace:
  [1] getindex(A::Vector{Union{}}, i1::Int64)
    @ Base ./array.jl:924
  [2] first(a::Vector{Union{}})
    @ Base ./abstractarray.jl:404
  [3] findnext(testf::DimensionalData.Dimensions.var"#11#12"{DataType, DataType}, A::Tuple{}, start::Int64)
    @ Base ./array.jl:2019
  [4] #s36#10
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:93 [inlined]
  [5] var"#s36#10"(::Any, f::Any, tosort::Any, order::Any)
    @ DimensionalData.Dimensions ./none:0
  [6] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any})
    @ Core ./boot.jl:582
  [7] _sortdims
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:84 [inlined]
  [8] _hasdim(f::Function, dims::Tuple{X{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, X{Colon}}}, Y{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, Y{Colon}}}, Ti{Sampled{DateTime360Day, Vector{DateTime360Day}, ForwardOrdered, Explicit{Matrix{DateTime360Day}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}}}}, query::Tuple{Val{Z}})
    @ DimensionalData.Dimensions ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:248
  [9] _call_primitive1
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:589 [inlined]
 [10] _call_primitive1
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:581 [inlined]
 [11] _call_primitive
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:576 [inlined]
 [12] _call_primitive
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:574 [inlined]
 [13] hasdim
    @ ~/.julia/packages/DimensionalData/K9D4P/src/Dimensions/primitives.jl:245 [inlined]
 [14] slice(x::Raster{Union{Missing, Float32}, 3, Tuple{X{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, X{Colon}}}, Y{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, Y{Colon}}}, Ti{Sampled{DateTime360Day, Vector{DateTime360Day}, ForwardOrdered, Explicit{Matrix{DateTime360Day}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}}}}, Tuple{}, Array{Union{Missing, Float32}, 3}, Symbol, Metadata{NCDfile, Dict{String, Any}}, Missing}, dims::Tuple{UnionAll})
    @ Rasters ~/.julia/dev/Rasters/src/methods/slice_combine.jl:19
 [15] slice(x::Raster{Union{Missing, Float32}, 3, Tuple{X{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, X{Colon}}}, Y{Mapped{Float64, Vector{Float64}, ForwardOrdered, Explicit{Matrix{Float64}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}, EPSG, EPSG, Y{Colon}}}, Ti{Sampled{DateTime360Day, Vector{DateTime360Day}, ForwardOrdered, Explicit{Matrix{DateTime360Day}}, Intervals{Center}, Metadata{NCDfile, Dict{String, Any}}}}}, Tuple{}, Array{Union{Missing, Float32}, 3}, Symbol, Metadata{NCDfile, Dict{String, Any}}, Missing}, dims::Type)
    @ Rasters ~/.julia/dev/Rasters/src/methods/slice_combine.jl:15
 [16] top-level scope
    @ REPL[61]:1
```

I have no clue why, since I didn't touch this method whatsoever (and neither none of the methods that are following).

**Comment by rafaqz** [comment]
Hmm ok, I can take a look over the weekend

**Comment by tcarion** [comment]
The things is that it works fine in the main branch so I really don't get it. Thanks, I'll have a closer look as well :) I didn't check if it would break rasters from other source yet

**Comment by tcarion** [comment]
Given where the tests fail, it seems `DimensionalData.hasdim` is also broken for Rasters from sources other than NetCDF.
After trying several things, it appears to start failing when importing `GRIBDatasets` (even with all the files looking like the main branch, it keeps failing until I comment out `using GRIBDatasets`). 
That's strange, because GRIBDatasets doesn't import DimensionalData, so I don't understand how it can break one of its methods. I checked if GRIBDatasets was committing type piracy, but I don't think so (all the extended methods from Base apply on types defined in GRIBDatasets). 
So the problem comes from GRIBDatasets, but I have no idea where it comes from exactly.

**Comment by tcarion** [comment]
Woops, I was actually committing type piracy :pirate_flag: !
I extended `Base.getindex` on `Tuple{Vararg{<:Dimension}}`, my bad. That was an elegant way of accessing a `Tuple` of `Dimension` with their name, but I will have to find another implementation

**Comment by rafaqz** [comment]
Ahh that explains it. Yeah, just use `dims(dimsyouhvae, dimyouwant)` 

Its kind of confusing how powerful `dims` is I should really make that clearer

**Comment by tcarion** [comment]
My previous message wasn't clear, I'm talking about `GRIBDatasets.Dimension`, not the one from `DD`.
Actually, it was no problem with DD or Rasters whatsoever, it was GRIBDatasets that was messing up the `Base.getindex` function the whole time. I extended it (along with a few other functions) on a `Tuple{Vararg{<:GRIBDatasets.Dimension}}`, thinking it would not mess with any other argument types because of the `<:GRIBDatasets.Dimension`, but it does :relaxed:. I still don't precisely understand why, but in any case, when I remove the lines where I extend those `Base` functions in GRIBDatasets, Rasters doesn't fail anymore (well of course it fails for grib sources since I broke `GRIBDatasets` :grin: ). Now I need to find another way of implementing the `Dimension`s in GRIBDatasets

**Comment by rafaqz** [comment]
Oh right! You are defining `getindex` for all empty Tuples with that vararg 🙃

**Comment by tcarion** [comment]
I was going on a quite painful (yet more robust) new implementation, but your fix just do the trick, thank you!

File: src/sources/grib.jl [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    end |> Tuple
+end
+
+DD.layermetadata(ds::GDS.Dataset) = _layermetadata(ds, Tuple(layerkeys(ds)))
+function _layermetadata(ds::GDS.Dataset, keys)
+    dimtypes = map(k -> DD.metadata(ds[string(k)]), keys)
+    NamedTuple{map(Symbol, keys)}(dimtypes)
+end
+
+missingval(var::GDS.Variable) = missing
```
Comment (rafaqz) [inline src/sources/grib.jl:82]: Are you actually converting nodata values to `missing` in GRIBDatasets.jl ?

**Review by rafaqz — COMMENTED** [review]
Looling good! A few minor comments if you need any feedback

File: src/sources/grib.jl (hunk 6 above)
Comment (tcarion) [reply src/sources/grib.jl:82]: Yes, I am

File: src/sources/common.jl [hunk 7]
```diff
@@ -0,0 +1,69 @@
+# Common utils methods for building Netcdf and GRIB Rasters.
+const NCD = NCDatasets
+const GDS = GRIBDatasets
+
+_unuseddimerror(dimname) = error("Netcdf contains unused dimension $dimname")
```
[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/common.jl:5]: true, thanks!

File: src/sources/grib.jl (hunk 6 above)
Comment (rafaqz) [reply src/sources/grib.jl:82]: I meant the commented out bits

File: src/sources/grib.jl [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+end
+
+function _dslookup(
+    ds::GDS.Dataset, dimname, D, index::AbstractArray{<:Union{Number,Dates.AbstractTime}},
+    metadata, crs, mappedcrs
+)
+    # Assume the locus is at the center of the cell if boundaries aren't provided.
+    # http://cfconventions.org/cf-conventions/cf-conventions.html#cell-boundaries
+    order = LA.orderof(index)
+    # var = NCD.variable(ds, dimname)
```
[the remark this answers is not in the document]
Comment (rafaqz) [reply src/sources/grib.jl]: I meant the commented out NCD bits

[the remark this answers is not in the document]
Comment (tcarion) [reply src/sources/grib.jl]: of course sorry, had a long day

**Comment by rafaqz** [comment]
@tcarion can you squash the last 88 commits and rebase or just start the branch/a new branch + PR again from main?

There seems to be some merge/rebase error that means we cant see the diff anymore, it's just all the changes made to the package since you started.

**Comment by tcarion** [comment]
Yes, I messed up with the rebase, sorry about that.
There have been some changes with GRIBDatasets and NCDatasets. They now both implement the [CommonDataModel](https://github.com/JuliaGeo/CommonDataModel.jl) interface, which should induce a smoother integration into Rasters and avoid duplicate code.
I'm already working on this, and a PR will follow soon. Afterwards, I suggest we close this one without merging.

**Comment by rafaqz** [comment]
Ahh ok great. Just make sure not to lose DiskArrays.jl integration in the update ;)

**Comment by rafaqz** [comment]
Also to clarify, I was assuming both CF standards and DiskArrays.jl would work with this source, so I didn't mean just keeping DiskArrays, but keeping it on the assumption that CF standards also worked. 

The netcdf source uses cf standards so they should also work here to be consistent.
